### ISSUE TITLE
[Feature Request] monaco-esbuild-plugin

### ISSUE THREAD

**Issue by ckorherr** [microsoft/monaco-editor#4614]
### Context

- [X] This issue is not a bug report. *(please use a different template for reporting a bug)*
- [X] This issue is not a duplicate of an existing issue. *(please use the [search](https://github.com/microsoft/monaco-editor/issues) to find existing issues)*

### Description

Monaco Editor integration with webpack is straightforward, allowing developers to easily configure the languages and features to include. With frameworks like Angular transitioning to esbuild, a similar plugin for esbuild would be great.

We developed our own esbuild plugin, which could serve as a starting point:

```
import * as path from 'path';
import * as fs from 'fs';
import * as crypto from 'crypto';
import { Plugin, PluginBuild } from 'esbuild';

const EDITOR_MODULE: IFeatureDefinition = {
  label: 'editorWorkerService',
  entry: undefined,
  worker: {
    id: 'vs/editor/editor',
    entry: 'vs/editor/editor.worker'
  }
};

interface IWorkerDefinition {
  id: string;
  entry: string;
}

interface IFeatureDefinition {
  label: string;
  entry: string | string[] | undefined;
  worker?: IWorkerDefinition;
}

interface ILoaderOptions {
  globals?: { [key: string]: string };
  pre?: string[];
  post?: string[];
}

interface IMonacoEditorEsbuildPluginOpts {
  languages?: string[];
  customLanguages?: IFeatureDefinition[];
  features?: string[];
  filename?: string;
  monacoEditorPath?: string;
  globalAPI?: boolean;
}

interface ILabeledWorkerDefinition {
  label: string;
  id: string;
  entry: string;
}

const resolveMonacoPath = (filePath: string, monacoEditorPath: string | undefined): string => {
  if (monacoEditorPath) {
    return require.resolve(path.join(monacoEditorPath, 'esm', filePath));
  }

  try {
    return require.resolve(path.join('monaco-editor/esm', filePath));
  } catch (err) {}

  try {
    return require.resolve(path.join(process.cwd(), 'node_modules/monaco-editor/esm', filePath));
  } catch (err) {}

  return require.resolve(filePath);
};

const getWorkerFilename = (filename: string, entry: string, monacoEditorPath: string | undefined): string => {
  const content = fs.readFileSync(resolveMonacoPath(entry, monacoEditorPath));
  const hash = crypto.createHash('md5').update(content).digest('hex').slice(0, 8);
  return filename.replace(/\[name\]/g, path.basename(entry, path.extname(entry))).replace(/\[contenthash\]/g, hash);
};

const getEditorMetadata = (monacoEditorPath: string | undefined): any => {
  const metadataPath = resolveMonacoPath('metadata.js', monacoEditorPath);
  return require(metadataPath);
};

const resolveDesiredFeatures = (
  metadata: any,
  userFeatures: string[] | undefined
): IFeatureDefinition[] => {
  const featuresById: { [feature: string]: IFeatureDefinition } = {};
  metadata.features.forEach((feature: IFeatureDefinition) => (featuresById[feature.label] = feature));

  function notContainedIn(arr: string[]) {
    return (element: string) => arr.indexOf(element) === -1;
  }

  let featuresIds: string[];

  if (userFeatures && userFeatures.length) {
    const excludedFeatures = userFeatures.filter((f) => f[0] === '!').map((f) => f.slice(1));
    if (excludedFeatures.length) {
      featuresIds = Object.keys(featuresById).filter(notContainedIn(excludedFeatures));
    } else {
      featuresIds = userFeatures;
    }
  } else {
    featuresIds = Object.keys(featuresById);
  }

  return coalesce(featuresIds.map((id) => featuresById[id]));
};

const resolveDesiredLanguages = (
  metadata: any,
  userLanguages: string[] | undefined,
  userCustomLanguages: IFeatureDefinition[] | undefined
): IFeatureDefinition[] => {
  const languagesById: { [language: string]: IFeatureDefinition } = {};
  metadata.languages.forEach((language: IFeatureDefinition) => (languagesById[language.label] = language));

  const languages = userLanguages || Object.keys(languagesById);
  return coalesce(languages.map((id) => languagesById[id])).concat(userCustomLanguages || []);
};

const coalesce = <T>(array: ReadonlyArray<T | undefined | null>): T[] => {
  return array.filter(Boolean) as T[];
};

const flatArr = <T>(items: (T | T[])[]): T[] => {
  return items.reduce((acc: T[], item: T | T[]) => {
    if (Array.isArray(item)) {
      return acc.concat(item);
    }
    return acc.concat([item]);
  }, []);
};

const fromPairs = <T>(values: [string, T][]): { [key: string]: T } => {
  return values.reduce(
    (acc, [key, value]) => Object.assign(acc, { [key]: value }),
    {} as { [key: string]: T }
  );
};

const monacoEditorEsbuildPlugin = (options: IMonacoEditorEsbuildPluginOpts = {}): Plugin => {
  const monacoEditorPath = options.monacoEditorPath;
  const metadata = getEditorMetadata(monacoEditorPath);
  const languages = resolveDesiredLanguages(metadata, options.languages, options.customLanguages);
  const features = resolveDesiredFeatures(metadata, options.features);
  const filename = options.filename || '[name].worker.js';
  const globalAPI = options.globalAPI || false;

  return {
    name: 'monaco-editor-esbuild-plugin',
    setup(build) {
      const modules = [EDITOR_MODULE].concat(languages).concat(features);
      const workers: ILabeledWorkerDefinition[] = [];
      modules.forEach((module) => {
        if (module.worker) {
          workers.push({
            label: module.label,
            id: module.worker.id,
            entry: module.worker.entry,
          });
        }
      });

      // Handle worker files
      const workerEntries = workers.reduce((acc, w) => {
        acc[w.entry] = resolveMonacoPath(w.entry, undefined);
        return acc;
      }, {} as Record<string, string>);

      const workerBuildResult = buildWorkers(build, workerEntries);
      const namesWithHash = fromPairs(Object.entries(workerBuildResult.metafile.outputs).map(([key, value]) => {
        return [path.parse(value.entryPoint).name, key];
      }));

      const workerPaths = fromPairs(
        workers.map(({ label, entry }) => [label, getWorkerFilename(filename, entry, monacoEditorPath)])
      );
      if (workerPaths['typescript']) {
        // javascript shares the same worker
        workerPaths['javascript'] = workerPaths['typescript'];
      }
      if (workerPaths['css']) {
        // scss and less share the same worker
        workerPaths['less'] = workerPaths['css'];
        workerPaths['scss'] = workerPaths['css'];
      }
    
      if (workerPaths['html']) {
        // handlebars, razor and html share the same worker
        workerPaths['handlebars'] = workerPaths['html'];
        workerPaths['razor'] = workerPaths['html'];
      }
      Object.entries(workerPaths).forEach(([key, value]) => {
        workerPaths[key] = namesWithHash[path.parse(value).name];
      });

      build.onEnd((result) => {
        const { outputFiles, metafile } = workerBuildResult;
        
        if (outputFiles?.length) {
          result.outputFiles?.push(...outputFiles);
        }
        if (result.metafile && metafile) {
          Object.assign(result.metafile.inputs, metafile.inputs);
          Object.assign(result.metafile.outputs, metafile.outputs);
        }
      });

      // Handle monaco editor imports
      build.onResolve({ filter: /esm[/\\]vs[/\\]editor[/\\]editor.(api|main)/ }, (args) => {
        return {
          path: args.path,
          namespace: 'monaco',
        };
      });

      build.onLoad({ filter: /.*/, namespace: 'monaco' }, async (args) => {
        const stringifyRequest = (request: string) => {
          let relativePath = path.relative('node_modules/monaco-editor/esm/vs/editor', request);
          
          // Normalize Windows paths
          if (path.sep === '\\') {
            relativePath = relativePath.replace(/\\/g, '/');
          }
          
          if (!relativePath.startsWith('.')) {
            relativePath = './' + relativePath;
          }
        
          return JSON.stringify(relativePath);
        };

        const contents = `
self["MonacoEnvironment"] = (function (paths) {
  return {
    globalAPI: ${globalAPI},
    getWorkerUrl: function (moduleId, label) {
      return paths[label];
    }
  };
})(${JSON.stringify(workerPaths, null, 2)});

${flatArr(coalesce(features.map((feature) => feature.entry))).map((entry) => `import ${stringifyRequest(resolveMonacoPath(entry, monacoEditorPath))};`).join('\n')}

import * as monaco from "./editor.api.js";
export * from "./editor.api.js";
export default monaco;

${flatArr(coalesce(languages.map((language) => language.entry))).map((entry) => `import ${stringifyRequest(resolveMonacoPath(entry, monacoEditorPath))};`).join('\n')}

export { WorkerManager } from "../language/typescript/tsMode";
export { CommandsRegistry } from '../platform/commands/common/commands';
        `;

        return {
          contents,
          loader: 'js',
          resolveDir: 'node_modules/monaco-editor/esm/vs/editor',
        };
      });
    },
  };
};

function buildWorkers(build: PluginBuild, entryPoints: Record<string, string>) {
  return build.esbuild.buildSync({
    platform: 'browser',
    write: false,
    bundle: true,
    metafile: true,
    outdir: '.',
    format: 'iife',
    entryPoints,
    loader: { '.ttf': 'file', '.css': 'css' },
    sourcemap: false,
    supported: undefined,
    plugins: undefined,
    splitting: false,
    entryNames: '[name]-[hash]',
  });
}

const monacoPlugin = monacoEditorEsbuildPlugin({
  languages: ['javascript', 'json', 'typescript', 'html', 'css', 'markdown'],
});

export default monacoPlugin;
```

For Angular projects, this plugin can be used with the [custom-esbuild](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-esbuild) builder.

With this approach we had to export WorkerManager and CommandsRegistry for registering a custom language.

**Comment by chriswnewman** [comment]
@ckorherr have yall come up with a solution to the warnings / errors related to web workers? ![Image](https://github.com/user-attachments/assets/1ab26850-1962-41c3-81a1-205cae2ba782)

your plugin looks promising, but I've ended up in the same spot from when I tried following this thread: https://github.com/microsoft/monaco-editor/issues/2411 (stuck on the above warnings/errors)

**Comment by omar-azmi** [comment]
@chriswnewman I faced a similar issue, and I found that the culprit is the use of relative paths when instantiating a new `Worker`.
This error originates from the `getWorker` function in `monaco-editor/esm/vs/base/browser/defaultWorkerFactory.js`.
To resolve this, you'd need to provide an absolute `http://` path to your worker.

What I do is that I set `window.MonacoEnvironment.getWorkerUrl` to use `import.meta.resolve(relative_path_to_worker)` to return an absolute path.

here is a little snippet of my `MonacoEnvironment.getWorkerUrl` logic:

```ts
// file "./load_workers.ts"

import { editor as monacoEditor, type Environment } from "monaco-editor"

declare const
	editorWorkerUrl: string,
	cssWorkerUrl: string,
	htmlWorkerUrl: string,
	jsonWorkerUrl: string,
	tsWorkerUrl: string

const global_context = globalThis as (typeof globalThis & { MonacoEnvironment?: Environment })
const monacoEnvironment: Environment = (global_context.MonacoEnvironment ??= {})
if (!monacoEnvironment.getWorkerUrl) {
	monacoEnvironment.getWorkerUrl = (worker_id: string, label: string) => (import.meta.resolve(editorWorkerUrl))
}
const original_getWorkerUrl = monacoEnvironment.getWorkerUrl!
monacoEnvironment.getWorkerUrl = (worker_id: string, label: string): string => {
	const subpath = workerSubpathResolver(worker_id, label)
	if (subpath) { return import.meta.resolve(subpath) }
	return original_getWorkerUrl(worker_id, label)
}

const workerSubpathResolver = (worker_id: string, label: string): string | undefined => {
	switch (label) {
		case "json":
			return jsonWorkerUrl
		case "css":
		case "scss":
		case "less":
			return cssWorkerUrl
		case "html":
		case "handlebars":
		case "razor":
			return htmlWorkerUrl
		case "typescript":
		case "javascript":
			return tsWorkerUrl
		default:
			return undefined
	}
}

export {
	monacoEditor,
	monacoEnvironment
}
```

and then in my `index.ts` or `index.html` file, I import `monacoEditor` from the worker loader file:

```ts
// file "./index.ts"

import { monacoEditor } from "./monaco_loader.ts"
import "./style.css"

export const editor = monacoEditor.create(document.getElementById("monaco-editor")!, {
	value: [
		"const a = 42",
		"const b: string = `Hello World!`",
		"",
	].join("\n"),
	language: "typescript",
	theme: "vs-dark",
})
```

**Comment by omar-azmi** [comment]
By the way, I have written a simple esbuild plugin that bundles this editor for the web, without the client having to hard-code any paths to the workers at all.
I was wondering if I could contribute a PR for it, so that there's an official esbuild plugin, similar to how there's an official webpack plugin.
If so, that would be awesome! But I'll need some guidance in terms of the stylistic/preferential choices listed below:
- what directory should the plugin reside in? (my initial assumption would be [`/esbuild-plugin/`](./esbuild-plugin/), similar to how the webpack plugin resides in [`/webpack-plugin`](./webpack-plugin)).
- may I use typescript (prefered)? or should it be written in plain javascript with jsdocs?
- should I also submit a "guide" under [`/docs/integrate-esm.md`](./docs/integrate-esm.md) on how to use the plugin, in addition to how to write a simple DIY esbuild plugin for monaco yourself?
- should I also submit a sample package under [`/samples/browser-esm-esbuild-plugin/`](./samples/browser-esm-esbuild-plugin/) that demonstrates how to setup the build script to bundle your code?
- finally, should I submit one PR, or multiple PRs, each covering one task (like, a single PR for doc guides, and a different one for the `samples` package)?

Thanks
(someone let me know if I should specifically tag someone for this)

**Comment by techfg** [comment]
> By the way, I have written a simple esbuild plugin that bundles this editor for the web, without the client having to hard-code any paths to the workers at all.
> I was wondering if I could contribute a PR for it, so that there's an official esbuild plugin, similar to how there's an official webpack plugin.

@omar-azmi - Would love to see a PR for this myself, the monaco-editor/esbuild story isn't great at the moment and I suspect your plugin would greatly improve it!  Do you have a fork of the plugin?  I'd be happy to do some testing of it, thank you!

**Comment by hediet** [comment]
Unfortunately, we currently do not have the resources to do lots of feature work on the monaco editor package.
However, personally, I would love to see a future where the monaco editor esm build is directly consumable by most bundlers, without any additional configuration or plugins (and thus eventually deprecating/removing the webpack plugin and not creating a monaco-editor specific esbuild plugin).

So far, the web workers are the main blocker, as there does not seem to be a _bundler-stable_ way to reference a web worker module. CSS imports are also problematic, as they cannot be run from source.
(code is _bundler-stable_ if it works as expected when running from sources (after stripping ts types), but also after bundling it with most bundlers).

However, some bundlers support the pattern `new Worker(new URL("./worker.js", import.meta.url))`, which we could make use of.

As for language and feature configuration, I believe this should be done via tree-shaking and `import.meta.env` or manual module imports.
To get feature parity with the webpack plugin, we could support these flags (though `import.meta.env.` is not standard conform yet and cannot be configured when running from source):

* `import.meta.env.MONACO_EDITOR_INCLUDE_FEATURES` (default `1`)
* `import.meta.env.MONACO_EDITOR_INCLUDE_FEATURE_{N}` (default `undefined`, if set, overrides the `FEATURES` value)
* `import.meta.env.MONACO_EDITOR_INCLUDE_LANGUAGES` (default `1`)
* `import.meta.env.MONACO_EDITOR_INCLUDE_LANGUAGE_{N}` (default `undefined`, if set, overrides the `LANGUAGES` value)

**Comment by omar-azmi** [comment]
@techfg Here is sample project that I drafted which uses a custom esbuild worker plugin for bunding:
- [omar-azmi/experiments_ts/tree/main/monaco_esbuild4](https://github.com/omar-azmi/experiments_ts/tree/main/monaco_esbuild4)

If you clone it, and execute `npm run build`, followed by `npm run serve`, you'll get a monaco-editor hosted onto your `http://localhost:8000/`, with the typescript-lsp worker running.

The plugin inside of [`./build.ts`](https://github.com/omar-azmi/experiments_ts/blob/main/monaco_esbuild4/build.ts) is fairly simple, and essentially works for bundling any kind of js-worker file.

One drawback of the approach that I take in the plugin is that each worker file is bundled separately on its own, meaning that code-splitting among worker files is not possible.

I don't know if the webpack plugin supports code-splitting among worker files, but if it does, then implementing it would need some thinking (for instance, one hacky approach could be to wait 1000ms until all worker files have _hopefully_ been discovered, then bundle them all together with code-splitting, then read the virtual output files, and distribute them back to each loader that's awaiting for the workers to bundle up)

@hediet 
> So far, the web workers are the main blocker, as there does not seem to be a _bundler-stable_ way to reference a web worker module.

In esbuild, one could bundle references to files using the [file-loader](https://esbuild.github.io/content-types/#file).
The bundled code will then contain a variable by the same name as the orginal import symbol, and it will be assigned a string referencing the relative-path to the file. In addition, the referenced file will also get copied over to the `outdir`.

This is what I'm essentially doing in my plugin demo project to attain the path to the bundled js-worker files.
There are some doc comments on the [`./src/monaco_loader.ts`](https://github.com/omar-azmi/experiments_ts/blob/138a519b7e2ca65e4345306ca725b88f638e0002/monaco_esbuild4/src/monaco_loader.ts#L1-L28) file which demonstrate the transformation.

So, I personally think it is feasible to accomplish worker-referencing with esbuild.
Would you be interested in a PR for a plugin similar to the one in the demo project?

**Comment by omar-azmi** [comment]
Adding support for `new Worker(new URL("./worker.js", import.meta.url))` through the plugin layer is non-trivial, because you would need to:
- Intercept every single file in the resolver.
- Preload the file **in the resolver**, and scan for the `new Worker(new URL(...))` pattern.
- If instances of the said pattern are found, then pass it over to a custom loader function (via namespace).
- The custom loader will transform all instances of the worker-url pattern to either:
  1. ```ts
     import worker_url from "resolved/path/to/worker.js"
     const worker = new Worker(import.meta.resolve(worker_url))
     // this transformation will be kind of hard to embed into the script,
     // but it will produce a static path variable to the `worker_url`.
     ```
  2. ```ts
     new Worker((await import("resolved/path/to/worker.js")).default)
     // this transformation can be inlined with a regex replace,
     // however, it will require an async context in order to await.
     ```
- And finally, _another_ custom plugin will need to intercept all "worker.js" files, and then bundle them separately (i.e. similar to my prior demo plugin)
  You can see the output of this kind of process in this [esbuild repl page](https://esbuild.github.io/try/#YgAwLjI1LjEAewogIGJ1bmRsZTogdHJ1ZSwKICBvdXRkaXI6ICcuL291dC8nLAogIGZvcm1hdDogJ2VzbScsCiAgc3BsaXR0aW5nOiB0cnVlLAogIGxvYWRlcjogewogICAgIi53b3JrZXIuanMiOiAiZmlsZSIsCiAgICAiLndvcmtlci50cyI6ICJmaWxlIiwKICB9LAp9AGUAZW50cnkudHMAaW1wb3J0IHdvcmtlcl91cmwgZnJvbSAiLi9sc3AtMS53b3JrZXIudHMiCgpjb25zdCBzdGF0aWNfaW1wb3J0X3dvcmtlciA9IG5ldyBXb3JrZXIoaW1wb3J0Lm1ldGEucmVzb2x2ZSh3b3JrZXJfdXJsKSkKCmNvbnN0IGR5bmFtaWNfaW1wb3J0X3dvcmtlciA9IG5ldyBXb3JrZXIoKGF3YWl0IGltcG9ydCgiLi9sc3AtMi53b3JrZXIudHMiKSkuZGVmYXVsdCkAAGxzcC0xLndvcmtlci50cwBjb25zb2xlLmxvZygiW0xTUC0xXTogaGVsbG8gd29ybGQhIikAAGxzcC0yLndvcmtlci50cwBjb25zb2xlLmxvZygiW0xTUC0yXTogaGVsbG8gd29ybGQhIik)

So while it might be possible, I think it might break some interoperability with other esbuild plugins, since we would be intercepting every resource, and not to mention that it will double the number of filesystem reads.

Alternatively, we could wait for the https://github.com/evanw/esbuild/pull/2508 PR to merge, and provide us with native handling of URL assets. (although it has been 2 years since the last update on the PR).

**Comment by techfg** [comment]
@omar-azmi 

> [@techfg](https://github.com/techfg) Here is sample project that I drafted which uses a custom esbuild worker plugin for bunding:
> 
> * [omar-azmi/experiments_ts/tree/main/monaco_esbuild4](https://github.com/omar-azmi/experiments_ts/tree/main/monaco_esbuild4)

Thanks for sharing!  I was able to get everything up and running as you describe.  I found a couple of issues, creating https://github.com/omar-azmi/experiments_ts/issues/1 for one of them.  The other, I'm still trying to isolate further but in short, in my actual project, it seems that the main bundle is being built multiple times - will create an issue if/when I can narrow this down further.  I also created https://github.com/techfg/experiments_ts/tree/monaco_esbuild_react/monaco_esbuild4_react which has your plugin working with [@monaco-editor/react](https://github.com/suren-atoyan/monaco-react) in a very basic sample.

@hediet 

Appreciate you providing your thoughts and input, thank you!

> Unfortunately, we currently do not have the resources to do lots of feature work on the monaco editor package. However, personally, I would love to see a future where the monaco editor esm build is directly consumable by most bundlers, without any additional configuration or plugins (and thus eventually deprecating/removing the webpack plugin and not creating a monaco-editor specific esbuild plugin).

This would be ideal.  The main challenge that exists with monaco and esm, especially with esbuild which currently doesn't have OOB bundler/plugin solution, is that the AMD loader is used which then causes issues with other dependencies that do not yet fully support esm (e.g., https://github.com/farzher/fuzzysort/issues/94#issuecomment-2118850668).  The possible solutions for this type of situation is: 1) Don't use dependencies that do not fully support ESM (not really an option); 2) bundle monaco as ESM (not easy and very fragile); 3) ensure that `define.amd` is `undefined` when bundling (e.g., via [define](https://esbuild.github.io/api/#define) config option of esbuild) to ensure deps that do not support ESM avoid the AMD loader.

In lieu of a full monaco ESM solution, as you mention, possibly @omar-azmi plugin could become a community plugin outside of monaco-editor directly but it would be good to see an official plugin for esbuild in monaco-editor if at all possible since, given your insight on available resources for the editor, it may be quite a while until we see full ESM support.

**Comment by hediet** [comment]
@omar-azmi thanks for your input!

> Adding support for new Worker(new URL("./worker.js", import.meta.url)) through the plugin layer is non-trivial

[I explored a plugin for rollup/vite here](https://github.com/microsoft/vscode/compare/main...hediet/b/monaco-editor-wip#diff-251f8bf1fe138b4c420d251d76251a738301c43b2b1f67a0e266ff6dc6de2eb6) and it worked quite well in that example.
I can imagine a plugin for esbuild might be more complicated though, as you described.

Is there any other way to ship a UI library with webworkers in a bundler-stable/independent way without library specific plugins, so that adding a new webworker doesn't require every bundler-user to adopt that change?
Having a library specific plugin for each bundler is a maintenance nightmare, especially if the maintainers don't know the bundlers very well.

> it may be quite a while until we see full ESM support.

Since VS Code fully moved to ESM some months ago, we plan to make the monaco editor an ESM-first package (probably within the next 6-12 months), slowly deprecating the AMD support (e.g. by introducing a legacy AMD package that acts as a wrapper).

**Comment by omar-azmi** [comment]
@hediet Thank you for the links. It got me wondering if there was an easy way to inspect and modify all loaded code before esbuild begins its minification, and to my surprise someone had already written such a library: https://github.com/aleclarson/esbuild-extra .

I can't quite grasp how the library's author manages to intercept the moment something has _already_ been loaded, and then let the user transform it midway, however they want. But it really works like magic.

With that library, I was able to easily transform `new URL(..., import.meta.url)` and `import.meta.resolve(...)` statements to static imports that esbuild can now redirect to my previous worker-bundling plugin.
If you're interested in checking it out, it's on the same repo as before: [omar-azmi/experiments_ts/tree/main/monaco_esbuild4](https://github.com/omar-azmi/experiments_ts/tree/c10104e24b27126b93754e672b5ae6eebe69093f/monaco_esbuild4)

I've also restructured the code a little bit so that the two plugins (the url-transformer, and the worker-bundler) sit under the [`./plugins/`](https://github.com/omar-azmi/experiments_ts/tree/c10104e24b27126b93754e672b5ae6eebe69093f/monaco_esbuild4/plugins) directory.

There still is one big caveat with the worker-bundler plugin, and it's that if your worker file imports additional assets that cannot be bundled into a single file (such as sourcemaps or this [demo-worker](https://github.com/omar-azmi/experiments_ts/blob/c10104e24b27126b93754e672b5ae6eebe69093f/monaco_esbuild4/src/demo_worker.ts)'s relative text file reference), then you **must** make sure that your esbuild config's `assetNames` is explicitly defined not to use the `"[hash]"` label in it; otherwise the path references to the assets generated inside of the worker-bundler will break.

### tl;dr

It is possible to write esbuild plugins that can import js-workers using the any of following statements:

```ts
import editorWorkerUrl from "monaco-editor/esm/vs/editor/editor.worker.js" with { type: "monaco-worker" }
import cssWorkerUrl from "monaco-editor/esm/vs/language/css/css.worker.js?worker"
const htmlWorkerUrl = new URL("monaco-editor/esm/vs/language/html/html.worker.js", import.meta.url).toString()
const jsonWorkerUrl = new URL("../node_modules/monaco-editor/esm/vs/language/json/json.worker.js", import.meta.url).toString()
const tsWorkerUrl = import.meta.resolve("monaco-editor/esm/vs/language/typescript/ts.worker.js")
const htmlWorker = new Worker(import.meta.resolve("monaco-editor/esm/vs/language/html/html.worker.js"))
const demoWorkerUrl = new URL("./demo_worker.ts", import.meta.url).toString()
```
